### PR TITLE
Unify JSON handling across RDF and RDG

### PR DESCRIPTION
**Pull request by captainsafia**
RDG currently resolves the `JsonTypeInfo` when deserializing a request body parameter once per request. This  applies an update to the codegen so that the `JsonTypeInfo` is resolved once at delegate construction and then re-used for each request.

RDG's implementation for response writing is currently slightly different from what RDF uses. There's a work item to make a public API (https://github.com/dotnet/aspnetcore/issues/47973) here but for now we unify the two implementations.

### CODE REVIEW

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_JsonBodyOrService_HandlesBothJsonAndService.generated.txt
Comment (BrennanConroy) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_JsonBodyOrService_HandlesBothJsonAndService.generated.txt:258]: The `WriteAsJsonAsync` call a couple lines up could be passed the non `T` options as well

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointJsonPreparationEmitter.cs
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/Emitters/EndpointJsonPreparationEmitter.cs:43]: The nullability is only emitted from the type that we pass into the `typeof` call (`GetTypeInfo(typeof(Todo?))` vs `GetTypeInfo(typeof(Todo))`). The resolved type info still contains the correct annotation (`JsonTypeInfo<Todo?>`).

I didn't add any new tests since we already have coverage for body binding scenarios but [this test](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.JsonBody.cs#L79) covers reading from types with various annotations....

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.Responses.cs
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.Responses.cs:280]: Your change is regressing this scenario because we are now ALWAYS calling `GetTypeInfo(typeof(object))`? Is that correct? (this test is passing in main) Maybe we shouldn't be doing this change if it is going to regress scenarios. I just thought you were hitting issues in new tests you were converting (for filters).

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:297]:
```suggestion
```

These should no longer be needed. You only need suppressions on the method calling the "unsafe" method. This method now calls `WriteJsonResponseAsync` which isn't annotated as "unsafe".

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:137]:
```suggestion
                        var todo_resolveBodyResult = await GeneratedRouteBuilderExtensionsCore.TryResolveBodyAsync<global::Microsoft.AspNetCore.Http.Generators.Tests.Todo>(httpContext, logOrThrowExceptionHelper, false, "Todo", "todo", todo_JsonTypeInfo);
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:201]: (nit)
```suggestion
                    var objectJsonTypeInfo = (JsonTypeInfo<object?>)serializerOptions.GetTypeInfo(typeof(object));
                    var todo_JsonTypeInfo = (JsonTypeInfo<global::Microsoft.AspNetCore.Http.Generators.Tests.Todo?>)serializerOptions.GetTypeInfo(typeof(global::Microsoft.AspNetCore.Http.Generators.Tests.Todo));
```

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:200]: Why is this `JsonTypeInfo<object?>` and not `JsonTypeInfo<object>`?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:253]: What happens if the filter returns `null`? What gets written to the Response?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.Responses.cs
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.Responses.cs:280]: > Your change is regressing this scenario because we are now ALWAYS calling GetTypeInfo(typeof(object))? Is that correct?

Yes, although it's only regressing the scenario when a JsonContext is used.

> Maybe we shouldn't be doing this change if it is going to regress scenarios.

I think we should still do it, but I'm comfortable waiting until the bug fix in the runtime is merged to merge this in to our repo.

> I just thought you were hitting issues in new tests you were converting (for filters).

The filters scenario was to showcase that the bug was happening in RDF as well and wasn't a side-effect of RDG's codegen, but as we eventually realized, any invocation of `GetTypeInfo(typeof(object))` will trigger this at the moment.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt]: Good spot -- this is not correct.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:253]: The response remains untouched. This matches the behavior of the filter construction pipeline that the runtime uses ([ref](https://github.com/dotnet/aspnetcore/blob/6575b289081b9f85fb30d01c7b306d1ec5d00a68/src/Http/Routing/src/RequestDelegateFilterPipelineBuilder.cs#L56-L61)).

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_BindAsync_Snapshot.generated.txt
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_BindAsync_Snapshot.generated.txt:400]:
```suggestion
                    var jsonOptions = serviceProvider?.GetService<IOptions<JsonOptions>>()?.Value ?? new JsonOptions();
                    var objectJsonTypeInfo = (JsonTypeInfo<object>)jsonOptions.SerializerOptions.GetTypeInfo(typeof(object));
```

small duplication reduction.

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (eerhardt) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt:319]:
```suggestion
                return httpContext.Response.WriteAsJsonAsync(value, jsonTypeInfo);
```

This seems wrong because when `value` is null, we get here.

Maybe the right fix is to take a `JsonTypeInfo<T?> jsonTypeInfo`?

**Review by eerhardt — APPROVED** [review]
LGTM. Just a couple small nits/questions

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitBodyParam_ComplexReturn_Snapshot.generated.txt]: Yep, I tweaked the nullability in https://github.com/dotnet/aspnetcore/pull/48573/commits/a9b8d8ba5f49b597ce14f091f39abb7a4e4551ad

**Pull request merged**
